src/types/utils.d.mts

```diff
@@ -163,3 +157,191 @@ type GetDataConfigOption = GetDataConfig extends {
+ * it is difficult to call the function in a safe way. This means its uses are
+ * mostly niche and it should be avoided.
+ *
+ * Make sure you have a good reason to use this type. It is almost always better
+ * to use a more specific function type. Please consider leaving a comment about
+ * why this type is necessary.
+ *
+ * Use this instead of:
+ * - `Function` - This refers to the fundamental `Function` object in JS. It allows classes.
+ * - `(...args: any[]) => any` - If someone explicitly accesses the parameters or return value you get `any` which is not safe.
+ * - `(...args: unknown[]) => unknown` - This allows obviously unsound calls like `fn(1, "foo")` because it indicates it can take any arguments.
+ */
+export type AnyFunction = (...args: never[]) => unknown;
+
+/**
+ * Use this type to allow any class, abstract class, or class-like constructor.
+ *
+ * See {@link AnyConcreteConstructor} if you cannot allow abstract classes. Please
+ * also consider writing a comment explaining why {@link AnyConcreteConstructor}
+ * is necessary.
+ *
+ * @example
+ * ```ts
+ * const concrete: AnyConstructor = class Concrete { ... }
+ * const abstract: AnyConstructor = abstract class Abstract { ... }
+ *
+ * // `Date` is not actually a class but it can be used as a constructor.
+ * const classLike: AnyConstructor = Date;
+ * ```
+ */
+export type AnyConstructor = abstract new (...args: never[]) => unknown;
+
+/**
+ * Use this type to allow any class or class-like constructor but disallow
+ * class-like constructors.
+ *
+ * Use this type only when abstract classes would be problematic such as the
+ * base type of a mixin. Please consider writing a comment explaining why.
+ * See {@link AnyConstructor} to also allow abstract classes.
+ *
+ * @example
+ * ```ts
+ * const concrete: AnyConcreteConstructor = class Concrete { ... }
+ *
+ * // `Date` is not actually a class but it can be used as a constructor.
+ * const classLike: AnyConcreteConstructor = Date;
+ *
+ * // This next line errors:
+ * const abstract: AnyConcreteConstructor = abstract class Abstract { ... }
+ * ```
+ */
+export type AnyConcreteConstructor = new (...args: never[]) => unknown;
+
+/**
+ * This type is equivalent to `Promise<T>`. Its usage is to indicate intent.
+ * Normally all callbacks should be of the form `(...) => MaybePromise<T>`
+ * because this gives the most flexibility to the caller. Typically asynchronous
+ * callbacks are simply awaited, meaning that there's no difference between
+ * `Promise` and {@link MaybePromise}. Even functions like
+ * {@link Promise.allSettled} function correctly with {@link MaybePromise}.
+ *
+ * Use this type only in the rare case where a type must be a `Promise`, for
+ * example if `promise.then` or `promise.catch` is explicitly called. Please
+ * also writing a comment explaining why {@link MaybePromise} is problematic in
+ * this context.
+ */
+export type MustBePromise<T> = Promise<T>;
```

Comment:
Should clarify intent for non-callback uses of Promise, e.g. various Document CRUD functions or Rolls that are natively async.